src/main/java/org/radarcns/management/web/rest/SourceResource.java

```diff
@@ -114,6 +119,38 @@
             return createSource(sourceDto);
         }
         checkPermission(getJWT(servletRequest), SOURCE_UPDATE);
+        Optional<SourceDTO> sourceToUpdateDto = sourceService
+                .findOneById(sourceDto.getId());
+
+        if (!sourceToUpdateDto.isPresent()) {
+            return ResponseEntity.notFound().headers(HeaderUtil.createFailureAlert(ENTITY_NAME,
+                "sourceNotFound",
+                "Cannot find a source by sourceName " + sourceDto.getSourceName())).build();
+        }
+        SourceDTO sourceToUpdate = sourceToUpdateDto.get();
+
+        // if the source is being transferred to another project.
+        if (!sourceToUpdate.getProject().getId().equals(sourceDto.getProject().getId())) {
+            if (sourceToUpdate.getAssigned()) {
+                return ResponseEntity.badRequest().headers(
+                    HeaderUtil.createFailureAlert(ENTITY_NAME, "sourceIsAssigned",
+                        "Cannot transfer an assigned source")).build();
+            }
+            // check whether source-type of the device is assigned to the new project
+            // to be transferred.
+            Optional<SourceType> sourceType = projectRepository
+                    .findSourceTypeByProjectIdAndSourceTypeId(sourceDto.getProject().getId(),
+                    sourceToUpdate.getSourceType().getId());
+            if  (!sourceType.isPresent()) {
+                return ResponseEntity.badRequest().headers(HeaderUtil
+                    .createFailureAlert(ENTITY_NAME,
+                    "invalidTransfer", "Cannot transfer a source to a project which doesn't "
+                        + "have compatible source-type")).build();
+            }
+            // set old source-type, ensures compatibility
+            sourceDto.setSourceType(sourceToUpdate.getSourceType());
+
+        }
```

Comment:
I think we should move all this logic into `SourceService`, **including** the call to `save()`. This would make the code more re-usable, plus it would make the whole update transactional.

Reply:
Indeed. That was my preferred method. But the problem is mainly due to the inconsistent way of error handling. We have two ways of error handling. One is by throwing exception and have the `ExceptionTranslator` to translate this into body. However, it deviates from the original method that was followed in the FrontEnd app. It expects headers with custom parameters to show proper messages to the user. So I decided to stick to responding with headers, because this action mainly involves user interaction and proper message on the front end seems essential. I didn't want to return `ResponseEntity` from `SourceService` either. It sounded far from `SourceService` logics. So I chose this way.

Reply:
Hmm, i thought the frontend could handle also error messages in the body, but i might be mistaken. If not we can modify the `ExceptionTranslator` to set the headers, and use exceptions everywhere.

Ok let's keep it here for now and revisit the error handling later, but i would feel safer with at least the `save()` being delegated to the service so it's transactional 😄

Reply:
Yes. That is what I thought. So I will push the exceptionHandler changes too. I was thinking of doing it separately. But same PR doesn't hurt.
Save is done through `SourceService` which is Transactional. 
`  SourceDTO result = sourceService.save(sourceDto);`

Reply:
I don't know why we have specific annotations for some methods in classes which are Transactional.

Reply:
> Save is done through SourceService which is Transactional. `SourceDTO result = sourceService.save(sourceDto);`

Sorry, of course, my bad!

> I don't know why we have specific annotations for some methods in classes which are Transactional.

Indeed, the method-level annotations can go away if there is a class-level one.